walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,112 @@
+    int endOfLastMeeting = 0;
+
+    // numberOfActiveMeetings will be 0  - before start and after end of the day,
+    // 1 - if current time slot is free for all mandatory attendees,
+    // >1 - in any other case.
+  
+    // numberOfActiveMeetings will increase if new time point under consideration is a start for a meeting, and decrease in other case.
+    for(TimePoint event : importantEvents) {
+      if (numberOfActiveMeetings == 1) {
+        // Check if there is enough time for requested meeting.
+        if (event.time() - endOfLastMeeting >= request.getDuration()) {
+          TimeRange freeSlot = TimeRange.fromStartDuration(endOfLastMeeting, event.time() - endOfLastMeeting);
+          possibleSlots.add(freeSlot);
+        }
+      }
+      
+      if (event.type() == 1) {
+        numberOfActiveMeetings += 1;
+      } else {
+        numberOfActiveMeetings -= 1;
+        endOfLastMeeting = event.time();
+      }
+    }
+    return possibleSlots;
+  }
+
+  /** Return list of time slots for current request for mandatory and if possible for optional attendees. */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Create a list of existing meetings with mandatory attendees.
+    List<TimePoint> importantEvents = new ArrayList<>();
+    // Create a list of existing meetings with optional attendees considered as mandatory.
+    List<TimePoint> optionalEvents = new ArrayList<>();
+  
+    // List of possible time slots for mandatory attendees.
+    Collection<TimeRange> possibleMandatorySlots = new ArrayList<>();
+    // List of possible time slots for optional attendees.
+    Collection<TimeRange> possibleOptionalSlots = new ArrayList<>();
+
+    // Add start and end of the day as time points.
+    TimePoint startOfTheDay = new TimePoint(0, 1, -1);
+    TimePoint endOfTheDay = new TimePoint(24 * 60, 2, -1);
+
+    importantEvents.add(startOfTheDay);
+    importantEvents.add(endOfTheDay);
+
+    optionalEvents.add(startOfTheDay);
+    optionalEvents.add(endOfTheDay);
+
+    for (Event event : events) {
+      boolean mandatoryAttendee = false;
+      // Pick events with mandatory attendees (fill list for optional as well).
+      for (String attendee: request.getAttendees()) {
+        if (event.getAttendees().contains(attendee)) {
+          TimePoint startOfEvent = new TimePoint(event.getWhen().start(), 1, -1);
+          importantEvents.add(startOfEvent);
+          optionalEvents.add(startOfEvent);
+          TimePoint endOfEvent = new TimePoint(event.getWhen().end(), 2, -1);
+          importantEvents.add(endOfEvent);
+          optionalEvents.add(endOfEvent);
+          mandatoryAttendee = true;
+          break;
+        }
+      }
+      if (mandatoryAttendee) {
+        continue;
+      }
+
+      // Pick events with optional attendees.
+      for (String attendee: request.getOptionalAttendees()) {
```

Comment:
optional: You might also be able to replace this loop with an existing function.